### ISSUE TITLE
http_relative_path parameter is not taken into account in keycloak_validator.rb

### ISSUE THREAD

**Issue by jonathanmarsaud** [treydock/puppet-module-keycloak#277]
Hi,

Since Keycloak Quarkus releases, they removed the context-path `/auth` accordingly to their [Release Announcement](https://www.keycloak.org/migration/migrating-to-quarkus), causing a change into URL users and apps are using.

We have a lot of Kubernetes clusters using OIDC with Keycloak that have the `/auth` URL into their configurations, in the client side (`~/.kube/config`), server side (`/etc/kubernetes/manifests/kube-apiserver`), and SSO Side (Google Cloud Console) and do not want to move from that path to avoid any changes (at least, in the client side to not disrupt users).

So I used the `http_relative_path` of this Puppet module, [accordingly to what Keycloak Documentation said about this parameter](https://www.keycloak.org/server/all-config?q=http-relative-path&f=build). The parameter itself works as intended, but `lib/puppet/util/keycloak_validator.rb` is not taking into account this parameter with its check, leading to this at every Puppet Agent run, even if all is working great actually:

```
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak within timeout window of 60 seconds; giving up.
Notice: Applied catalog in 64.87 seconds
```

In my case, as I configured `http_relative_path => '/auth'`, it should check `http://localhost:8080/auth/`.

Regards, and thanks for this module!

**Comment by treydock** [comment]
The fix for this will be released as v9.3.3 once Github Actions complete.

**Comment by jonathanmarsaud** [comment]
Hello,

I'm reopening this issue because I just tested the `v9.3.3` but it doesn't seem to work properly.

Here's my Puppet code for Keycloak:

```puppet
  class { 'keycloak':
    hostname           => $hostname,
    version            => $keycloak_version,
    proxy              => 'edge',
    http_relative_path => '/auth',
    db                 => 'mariadb',
    db_url_host        => $ip,
    db_url_port        => $dbport,
    db_url_database    => $db,
    db_username        => $user,
    db_password        => $password,
    manage_db_server   => false,
    manage_db          => false,
  }
```

And here is what happen on a perfectly fine Keycloak with this update:

```
# puppet agent -t
Info: Using environment 'dev'
Info: Retrieving pluginfacts
Notice: /File[/opt/puppetlabs/puppet/cache/facts.d]/mode: mode changed '0755' to '0775'
Info: Retrieving plugin
Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet/provider/keycloak_conn_validator/puppet_https.rb]/content: 
--- /opt/puppetlabs/puppet/cache/lib/puppet/provider/keycloak_conn_validator/puppet_https.rb	2023-03-01 16:12:10.467228387 +0000
+++ /tmp/puppet-file20230314-647290-idp47d	2023-03-14 10:22:59.520825710 +0000
@@ -67,6 +67,6 @@
   #
   # @api private
   def validator
-    @validator ||= Puppet::Util::KeycloakValidator.new(resource[:keycloak_server], resource[:keycloak_port], resource[:use_ssl], resource[:test_url])
+    @validator ||= Puppet::Util::KeycloakValidator.new(resource[:keycloak_server], resource[:keycloak_port], resource[:use_ssl], resource[:test_url], resource[:relative_path])
   end
 end

Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet/provider/keycloak_conn_validator/puppet_https.rb]/content: content changed '{sha256}5ebe05ce820c83e6b6eb7c086f7406c06895a800a04a5498f8844e5555036a0c' to '{sha256}52454705e50f8b664390111730797bf63d5f6c4ce5cb758e4715ff10a705e5d7'
Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet/type/keycloak_conn_validator.rb]/content: 
--- /opt/puppetlabs/puppet/cache/lib/puppet/type/keycloak_conn_validator.rb	2023-03-01 16:12:11.495317195 +0000
+++ /tmp/puppet-file20230314-647290-p6dro6	2023-03-14 10:22:59.640836101 +0000
@@ -32,7 +32,12 @@
 
   newparam(:test_url) do
     desc 'URL to use for testing if the Keycloak database is up'
-    defaultto '/auth/admin/serverinfo'
+    defaultto '/realms/master/.well-known/openid-configuration'
+  end
+
+  newparam(:relative_path) do
+    desc 'URL relative path that is used by Keycloak'
+    defaultto '/'
   end
 
   newparam(:timeout) do

Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet/type/keycloak_conn_validator.rb]/content: content changed '{sha256}db88c03dbb101abba3dd929f58f321f38f0e8017901211388f6e8b4a8b27be60' to '{sha256}e6f2000b23fc870289858d01852597f307c6643690ab2d3a51b4250e6ab929c0'
Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet/util/keycloak_validator.rb]/content: 
--- /opt/puppetlabs/puppet/cache/lib/puppet/util/keycloak_validator.rb	2023-03-01 16:12:12.239381469 +0000
+++ /tmp/puppet-file20230314-647290-ue43hy	2023-03-14 10:22:59.700841298 +0000
@@ -6,11 +6,12 @@
 class Puppet::Util::KeycloakValidator
   attr_reader :keycloak_server, :keycloak_port, :use_ssl, :test_path
 
-  def initialize(keycloak_server, keycloak_port, use_ssl = false, test_path = '/auth/admin/serverinfo')
+  def initialize(keycloak_server, keycloak_port, use_ssl = false, test_path = '/realms/master/.well-known/openid-configuration', relative_path = '/')
     @keycloak_server = keycloak_server
     @keycloak_port   = keycloak_port
     @use_ssl         = use_ssl
     @test_path       = test_path
+    @relative_path   = relative_path
   end
 
   # Utility method; attempts to make an http/https connection to the keycloak server.
@@ -22,10 +23,11 @@
     # All that we care about is that we are able to connect successfully via
     # http(s), so here we're simpling hitting a somewhat arbitrary low-impact URL
     # on the keycloak server.
+    path = "#{@relative_path}#{@test_path.sub(%r{^/}, '')}"
     http = Net::HTTP.new(@keycloak_server, @keycloak_port)
     http.use_ssl = @use_ssl
     http.verify_mode = OpenSSL::SSL::VERIFY_NONE
-    request = Net::HTTP::Get.new(@test_path)
+    request = Net::HTTP::Get.new(path)
     request.add_field('Accept', 'application/json')
     response = http.request(request)
 

Notice: /File[/opt/puppetlabs/puppet/cache/lib/puppet/util/keycloak_validator.rb]/content: content changed '{sha256}1f3d7b41bf59b48fb3e91adc15d687a3daa74562e5852b09cc13010e2605e941' to '{sha256}0c83932525ea28f1779c3872af6cf96feb4dc70ce415c60986b51c755fe2f077'
Info: Loading facts
Info: Caching catalog for <redacted>
Info: Applying configuration version '1678789388'
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak; sleeping 5 seconds before retry
Notice: Unable to connect to keycloak server (http://localhost:8080): [404] Not Found
Notice: Failed to connect to keycloak within timeout window of 60 seconds; giving up.
Notice: Applied catalog in 64.72 seconds
Info: Loading facts
Info: Uploading facts for <redacted> to <redacted>
```

Do you have any clues?
Thanks,
